### ISSUE TITLE
IContentSerializer implementation and/or example for .NET Core 3.0?

### ISSUE THREAD

**Issue by martincostello** [reactiveui/refit#653]
**Is your feature request related to a problem? Please describe.**

I'm currently maintaining branches of some applications using ASP.NET Core 2.2 + Refit against the ASP.NET Core 3.0 previews.

As the new built-in JSON (de)serialization for 3.0 is still in flux, I'm continuing to use Netwonsoft.Json as the (de)serializer using the new `Microsoft.AspNetCore.Mvc.NewtonsoftJson` NuGet package and Refit's `JsonContentSerializer` class.

With updating to 3.0 preview 4, I'm now having to st `AllowSynchronousIO` to `true` on both Kestrel and TestServer in order to prevent transient failures from synchronous buffer flushes due to the synchronous nature of the Newtonsoft.Json APIs.

This highlights a need to me to consider transitioning away to the new JSON APIs in System.Text.Json to remove the need to opt-in to synchronous IO in this way.

If not built-in to Refit itself, it would be useful to the community to provide an example of an `IContentSerializer` (and if/when merged `IContentSerializerWithCancellation`) implementation that uses the new JSON APIs to enable people to easily get their applications to "keep working" with minimal opt-ins for "discouraged" code usages, such as synchronous IO.

**Describe the solution you'd like**

Either a NuGet package or documentation on how to use System.Text.Json with Refit.

**Describe alternatives you've considered**

  * Custom `` implementation in application source that uses System.Text.Json.
  * Continuing to use Newtonsoft.Json with synchronous IO enabled.

**Describe suggestions on how to achieve the feature**

  1. An extension NuGet package for Refit shipping in the ASP.NET Core 3.0 timeline that contains an appropriate `IContentSerializer` implementation for System.Text.Json.
  1. Documentation/sample code for a `IContentSerializer` implementation using System.Text.Json.

**Comment by martincostello** [comment]
I had a quick play around by reverse-engineering the MVC input/output formatters and came up with the below.

It doesn't 100% work in my app's tests because of a bug (https://github.com/dotnet/corefx/pull/36637) that doesn't appear to have made the cut for preview 4, but it appears to work for simple cases:

```cs
public class SystemTestJsonContentSerializer : IContentSerializer
{
    public SystemTestJsonContentSerializer(MvcOptions options)
    {
        SerializerOptions = options.SerializerOptions;
    }

    private JsonSerializerOptions SerializerOptions { get; }

    public async Task<T> DeserializeAsync<T>(HttpContent content)
    {
        Stream utf8Json = await content.ReadAsStreamAsync();
        return await JsonSerializer.ReadAsync<T>(utf8Json, SerializerOptions);
    }

    public async Task<HttpContent> SerializeAsync<T>(T item)
    {
        var stream = new MemoryStream();

        try
        {
            await JsonSerializer.WriteAsync(item, stream, SerializerOptions);
            await stream.FlushAsync();

            var content = new StreamContent(stream);

            content.Headers.ContentType = new MediaTypeHeaderValue("application/json")
            {
                CharSet = Encoding.UTF8.WebName,
            };

            return content;
        }
        catch (Exception)
        {
            await stream.DisposeAsync();
            throw;
        }
    }
}
```

**Comment by wrathofodin** [comment]
Refit should move away from Json.NET to the new `System.Text.Json`

**Comment by benjaminhowarth1** [comment]
Given that's quite a new implementation, we'll need to check compatibility with .NET 4.6.1 & Standard 1.4 first, but I'd agree we should consider it.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by onovotny** [comment]
.NET Standard 1.4 is no longer relevant as .NET Core 1.0 and 1.1 are both out of support now. .NET Standard 2 is our minimum.

**Comment by onovotny** [comment]
Keep in mind that Json.NET is far more flexible than System.Text.Json. The latter is meant for raw perf but the former can adapt to more shapes.

If we go with Sys.Text.Json for the default, we may want to provide an optional package that provides a Json.Net serializer/deserializer as well.

**Comment by Gakk** [comment]
I already have project using System.Text.Json to (de)serialize REST API. Would love to implement Refit, but I would like to not depend on Newtonsoft.Json or change attributes to control serialization of property names.

**Comment by martincostello** [comment]
Resolved by #836
